Bio/Sequencing/Ace.py

```diff
@@ -298,197 +298,209 @@ def parse(handle):
 
     where each record is a Contig object.
     """
-    handle = iter(handle)
-
-    line = ""
-    while True:
-        # at beginning, skip the AS and look for first CO command
-        try:
-            while True:
-                if line.startswith("CO"):
-                    break
-                line = next(handle)
-        except StopIteration:
-            return
-
-        record = Contig(line)
-
-        for line in handle:
-            line = line.strip()
-            if not line:
-                break
-            record.sequence += line
-
-        for line in handle:
-            if line.strip():
-                break
-        if not line.startswith("BQ"):
-            raise ValueError("Failed to find BQ line")
-
-        for line in handle:
-            if not line.strip():
-                break
-            record.quality.extend(int(x) for x in line.split())
-
-        for line in handle:
-            if line.strip():
-                break
-
-        while True:
-            if not line.startswith("AF "):
-                break
-            record.af.append(af(line))
-            try:
-                line = next(handle)
-            except StopIteration:
-                raise ValueError("Unexpected end of AF block")
-
-        while True:
-            if line.strip():
-                break
-            try:
-                line = next(handle)
-            except StopIteration:
-                raise ValueError("Unexpected end of file")
+    try:
+        handle = open(source)
+    except TypeError:
+        handle = source
+        if handle.read(0) != "":
+            raise ValueError("Ace files must be opened in text mode.") from None
 
+    try:
```

Comment:
Is this entire ``try`` clause just for using ``finally``? No ``except``?

Reply:
Correct. Perhaps we should put this whole block in a separate function.